common/s2n-codec/src/decoder/checked_range.rs

```diff
@@ -4,6 +4,7 @@
 use core::fmt;
 
 #[derive(Copy, Clone)]
+#[cfg_attr(test, derive(bolero::TypeGenerator))]
```

Comment:
This struct has an invariant (current implicit) where `end >= start`. the `TypeGenerator` impl should enforce that. We should probably also add a `debug_assertion` in `CheckedRange::new`.

Note that this will also break the `checked_range_unsafe` feature. Fortunately, I don't think that's actually used anywhere so we can probably remove it in another PR.

Reply:
Would this impl enforce the invariant `end >= start`? 

```rust 
#[test]
impl bolero::TypeGenerator for CheckedRange {
    fn generate<D: bolero::Driver>(driver: &mut D) -> Option<Self> {
        let start = gen::<usize>().generate(driver)?;
        let end = gen::<usize>().generate(driver)?;
        if end >= start {
            Some(CheckedRange {
                start, 
                end
            })
        } else {
            None
        }
    }
}

```

Reply:
Should the `debug_assertion` trigger when the invariant is broken in the call? Something like: 

```rust 
pub(crate) fn new(start: usize, end: usize, original_ptr: *const u8) -> Self {
  debug_assertion!(end >= start, "end: {end} must be greater than or equal to start: {start}");
  ...
}
```

Reply:
You could also do

```rust
impl bolero::TypeGenerator for CheckedRange {
    fn generate<D: bolero::Driver>(driver: &mut D) -> Option<Self> {
        let start = gen::<usize>().generate(driver)?;
        let end = (start..).generate(driver)?;
        Some(CheckedRange {
            start, 
            end
        })
    }
}
```

> Should the debug_assertion trigger when the invariant is broken in the call?

Yes what you've got there looks great!